python/lsst/meas/algorithms/brightStarStamps.py

```diff
@@ -0,0 +1,280 @@
+
+    def __len__(self):
+        return len(self._starStamps)
+
+    def __getitem__(self, index):
+        return self._starStamps[index]
+
+    def append(self, item):
+        """Add an additional bright star stamp.
+
+        Parameters
+        ----------
+        item : `dict`
+            Mapping containing keys ``"starStamp"``, ``"gaiaGMag"``,
+            ``"gaiaId"``, and `"annularFlux"`.
+        """
+        self._starStamps.append(self._checkStarStamp(item))
+        return None
+
+    def getMaskedImages(self):
+        """Retrieve star images.
+
+        Returns
+        -------
+        maskedImages :
+            `list` [`lsst.afw.image.maskedImage.maskedImage.MaskedImageF`]
+        """
+        return [stamp["starStamp"] for stamp in self._starStamps]
+
+    def getMagnitudes(self):
+        """Retrieve Gaia G magnitudes for each star.
+
+        Returns
+        -------
+        gaiaGMags : `list` [`float`[
+        """
+        return [stamp["gaiaGMag"] for stamp in self._starStamps]
+
+    def getGaiaIds(self):
+        """Retrieve Gaia IDs for each star.
+
+        Returns
+        -------
+        gaiaIds : `list` [`str`]
+        """
+        return [str(stamp["gaiaId"]) for stamp in self._starStamps]
+
+    def getAnnularFluxes(self):
+        """Retrieve normalization factors for each star.
+
+        These are computed by integrating the flux in annulus centered on the
+        bright star, far enough from center to be beyond most severe ghosts and
+        saturation. The inner and outer radii that define the annulus can be
+        recovered from the metadata.
+
+        Returns
+        -------
+        annularFluxes : list[`float`]
+        """
+        return [stamp["annularFlux"] for stamp in self._starStamps]
+
+    def selectByMag(self, magMin=None, magMax=None):
+        """Return the subset of bright star stamps for objects with specified
+        magnitude cuts (in Gaia G).
+
+        Parameters
+        ----------
+        magMin : `float`, optional
+            Keep only stars fainter than this value.
+        magMax : `float`, optional
+            Keep only stars brighter than this value.
+        """
+        selector = np.array([True for _ in range(len(self))])
+        if magMin is not None:
+            selector *= np.array(self.getMagnitudes()) > magMin
+        if magMax is not None:
+            selector *= np.array(self.getMagnitudes()) < magMax
+        selIdx = np.where(selector)[0]
+        return BrightStarStamps([self._starStamps[j] for j in selIdx], metadata=self._metadata)
+
+    def getMetadata(self):
+        """Retrieve metadata associated with these `Defects`.
+
+        Returns
+        -------
+        meta : `lsst.daf.base.PropertyList`
+            Metadata. The returned `~lsst.daf.base.PropertyList` can be
+            modified by the caller and the changes will be written to
+            external files.
+        """
+        return self._metadata
```

Comment:
My reasoning for having it as a private member with an explicit getter, which might very well not be sound, is the following.

The metadata of the class absolutely needs to contain some "new" information (annular radii especially, but also the Gaia magnitudes and ids), specific to the bright star stamps. In many use cases, it might also contain the metadata from the visit from which the stamps were extracted. 

My intention was for this to be transparent to the user. When these objects are created (DM-25304), the user can just pass on the exposure metadata unchanged, along with the annular radii they set when processing the stamps, and be sure the metadata will be written along with the stamps and new extra information.

Conversely, I was hoping this design would strongly imply the user probably should not alter the metadata manually.

A very similar design is used for defects: https://github.com/lsst/meas_algorithms/blob/b97ba8bf064112940b133beb7b5d8f5936140fec/python/lsst/meas/algorithms/defects.py#L219

Reply:
Having the object contain new information or not makes no difference to the access pattern of the instance attribute, you could still have `.metadata` even if you mutate it. An attribute is just a flatter way to access data contained by the class. However your comment does bring up some things:
* Since you are mutating the metadata passed into the class, you really should create a copy of it in `__init__`, since you don't know if your changes will affect something down stream
* If you want the user to not modify the metadata, it is better to return a copy of your internal metadata object in a getter
* but if you really want to make it read only in a pythonic way, you should keep it as a private attribute `_metadata` and have a metadata property like I demoed above. This will prevent someone from reassigning the `.metadata` to anything. If it is important that someone not assign to the property set you should also return a copy in the property, so that they can use values but any changes they make do not affect your class.. That will allow your class to have an interface such that someone could write `brightStars.metadata['some_property']` while still keeping the safety you want.

Putting that together would look something like:
```python
class BrightStarStamps:
    def __init__(..., metadata, ...):
        ...
        self._metadata = metadata.copy()
        ...
    @property
    def metadata(self):
        return self._metadata.copy()
```
Note copy is a shallow copy, so the container is new, and the keys are new, but the values are not duplicated. The amount of copying for protection is up to you, but as I suspect this is not a rapidly accessed member speed is probably not a concern.